### ISSUE TITLE
[modules] Slice Vavr and create Java 9 conform modules

### ISSUE THREAD

**Issue by danieldietrich** [vavr-io/vavr#2057]
We described it already in the blog post [Fast Forward >> Vavr 1.0](http://blog.vavr.io/fast-forward-to-vavr-1-0).

![img_9507](https://user-images.githubusercontent.com/743833/29488184-f7f3f974-8505-11e7-8713-6c2aa5dc2609.PNG)

I think additionally

* PartialFunction is needed in `vavr-core`
* Each package might get an API class that provides API shortcuts (we slice the current io.vavr.API)

**Comment by danieldietrich** [comment]
Currently this sequence method takes my thoughts. We plan to slice/modularize Vavr (see #2057). Especially there's the plan to have two modules, vavr-core and vavr-collection.

Seq will be located in vavr-collection, but there are Option.sequence and Try.sequence that will be located in vavr-core. This will be a problem, because Seq is not visible in vavr-core.

I can imagine different scenarios to solve that problem:

## Keep the sequence methods where they are and change their signature

```java
// module vavr-core
interface Option<T> {
    // before it returned Option<Seq<T>>
    static <T> Option<Iterable<T>> sequence(Iterable<? extends Option<? extends T>> values) { ... }
}
```

The new return value could also be based on another standard Java collection-like type:

* `Option<java.util.List<T>>`
* `Option<java.util.Iterator<T>>`
* ...

This would make it more cumbersome to lift the result to Vavr again:

```java
// before
Option<Seq<Integer>> o1 = Option.sequence(List(Some(1), Some(2)));
Seq<String> r1 = o1.getOrElse(List()).map(Object::toString);

// after
Option<Iterable<Integer>> o2 = Option.sequence(List(Some(1), Some(2)));
Seq<String> r2 = o2.map(List::ofAll).getOrElse(List()).map(Object::toString);
```

I think this is not a good solution because it would complicate things for a technical, (more or less) internal reason.

## Fuse the modules vavr-core and vavr-collection

This would solve the technical issue of visibility but it would contradict our desire to select features by including only those modules that are needed.

## Move the affected static(!) methods to the other module

We could move the sequence methods to Seq.

```java
// before
Option<Seq<T>> a2 = Option.sequence(options);
Try<Seq<T>> a3 = Try.sequence(tries);

// after
Either<Seq<T>> a1 = Seq.sequence(eithers);
Option<Seq<T>> a2 = Seq.sequence(options);
Try<Seq<T>> a3 = Seq.sequence(tries);
```

But there are several problems:

### 1) The method signatures would be ambiguous because of type erasure (if they share the same name 'sequence')

```java
interface Seq<T> {
    // to Java, these are all sequence(Iterable<Object>)
    static <T> Option<Seq<T>> sequence(Iterable<? extends Option<? extends T>> options) { ... }
    static <T> Try<Seq<T>> sequence(Iterable<? extends Try<? extends T>> options) { ... }
}
```

This could be solved by choosing different names.

### 2) It would not scale well

If we add a type, say Either or Validation, we would have to add also dependencies to other modules. We do not want to do this because it would contradict our desire to choose features.

```java
interface Seq<T> {
    static <T> Option<Seq<T>> sequenceOption(Iterable<? extends Option<? extends T>> options) { ... }
    static <T> Try<Seq<T>> sequenceTry(Iterable<? extends Try<? extends T>> options) { ... }
    static <T> Either<Seq<T>> sequenceEither(Iterable<? extends Either<? extends T>> eithers) { ... }
    static <T> Validation<Seq<T>> sequenceValidation(Iterable<? extends Validation<? extends T>> validations) { ... }
}
```

Generally not a good idea.

If we had higher-kinded types, we could preserve the type:

```java
```

---

Summing up,

* we want to have the sequence method

**Comment by danieldietrich** [comment]
One more word to the options we have regarding sequencing in the presence of Java modules.

## Higher-Kinded Types

I've slept one night over the topic and still think that simulated higher-kinded types are a **no-go**  for Vavr.

<details>

The dependency problem mentioned [above](https://github.com/vavr-io/vavr/issues/2057#issuecomment-328077343) would be solved, if we could define a _generic_ method `M<Seq<T>> Seq.sequence(Iterable<M<T>> values)` that is applicable to all monadic types M (like Option, Try, Either, ...).

Because of the lack of [higher-kinded types](https://typelevel.org/blog/2016/08/21/hkts-moving-forward.html) (aka type constructors), we  would have to 'simulate' them like this:

```java
interface Kind<K extends Kind<K, ?>, T> {}

interface Option<T> extends Value<T>, Kind<Option<?>, T> { ... }

interface Seq<T> extends Traversable<T> {
    @SafeVarargs
    static <T, M extends Kind<M, ?> & Value<?>, V extends Kind<M, T>>
        Kind<M, Seq<T>> sequence(V... values) { ... }
    static <T, M extends Kind<M, ?> & Value<?>, V extends Kind<M, T>>
        Kind<M, Seq<T>> sequence(Iterable<? extends V> values) { ... }
}
```

**Usage:**

```java
// Option<Seq<String>> is split into Kind<Option<?>, Seq<String>>
Kind<Option<?>, Seq<String>> seqOpt = Seq.sequence(Option.some("a"), Option.some("b");
```

This Kind type isn't useful for further processing. We need to re-union the split type. This isn't possible in a generic way (otherwise we would already have higher-kinded types in Java). However, we could achieve it by adding a static `narrow()` method to the corresponding type:

```java
interface Option<T>  extends Value<T>, Kind<Option<?>, T> {
    static <T> Option<T> narrow(Kind<Option<?>, T> kind) {
        return (Option<T>) kind;
    }

}
```

**Fixed Usage:**

```java
Option<Seq<String>> seqOpt = Option.narrow(Seq.sequence(Option.some("a"), Option.some("b"));
```

**Disadvantages of this solution:**

* library developers have to implement narrow() by convention because it is static
* the user has to remember that a narrow() call fixes the uncommon type problems
* we introduce a new type Kind that is part of the public API for technical reasons only
* the method signatures of our public API aren't human-readable any more - we want to keep it as simple as possible

Summing up, this is not the kind of API we want to present our users.

</details>

## Alternate Solution

We need to keep the sequence methods where they are, as static members at their corresponding types. The problem remains - the `Seq` type isn't known anymore after we sliced Vavr into modules. I see the following solution:

```java
interface Option<T> extends Value<T> {
    @SafeVarargs
    static <T> Option<T[]> sequence(Option<? extends T>... options) { ... }
    static <T> Option<T[]> sequence(Iterable<? extends Option<? extends T>> options) { ... }
}
```

**Usage:**

```java
// Option<String[]>
Option.sequence(Option.some("a"), Option.some("b"));

// Option<java.util.List<String>>
Option.sequence(Option.some("a"), Option.some("b")).map(Arrays::of);

// Option<Vector<String>>
Option.sequence(Option.some("a"), Option.some("b")).map(Vector::of);
```

**Trade-offs:**

* We need an additional map() call in order to transform the result to an appropriate type.
* Java does not allow 'unsound' type assignments, e.g.

```java
// fails to compile because Option<Vector<T>> is not of type Option<Seq<T>>
Option<Seq<T>> seqOpt = Option.sequence(Option.some("a"), Option.some("b")).map(Vector::of);
```

The type-case problem could be solved by offering new collection factory methods, e.g.

```java
interface Seq<T> extends Traversable<T> {
    static <T> Seq<T> of(T... ts) { return Vector.of(ts); }
}

// Usage
Option<Seq<T>> seqOpt = Option.sequence(Option.some("a"), Option.some("b")).map(Seq::of);
```

However, in most cases it should be ok to work with the more specific type, e.g. `Option<Vector<T>>`.

**Comment by zyxist** [comment]
Great to see another project that considers using modules :). Consider also using the recommended naming for modules:

 * Module names must be reverse-DNS, just like package names, e.g. org.joda.time.
 * Modules are a group of packages. As such, the module name must be related to the package names.
 * Module names are strongly recommended to be the same as the name of the super-package.
 * Creating a module with a particular name takes ownership of that package name and everything beneath it.
 * As the owner of that namespace, any sub-packages may be grouped into sub-modules as desired so long as no package is in two modules.

Using reverse-DNS style is recommended by Mark Reinhold, chief Java architect, and there is a good explanation written by Stephen Colebourne, one of Java experts:

 * http://blog.joda.org/2017/04/java-se-9-jpms-module-naming.html
 * https://github.com/jodastephen/jpms-module-names

**Comment by danieldietrich** [comment]
@zyxist thank you for your feedback! Yes, we will use reverse-DNS style and name the modules like the super packages. No package will occur twice (in different modules).

**Comment by hamnis** [comment]
I think Either should be in core.

**Comment by danieldietrich** [comment]
I would prefer to move Option, Try, Either and Validation out of `io.vavr.core` to a module `io.vavr.control` to get rid of the module `io.vavr.controlx`.

The crux is that Function*'s lift and liftTry currently use Option and Try, so the core depends on these types. Now that we have PartialFunction it might be a viable solution to get rid of Function*.lift and .liftTry.

**Comment by danieldietrich** [comment]
The basic problem of modularization is understood and solved for Vavr. We will not incrementally add more modules. The first module on current master is vavr-control.
